pkg/distributor/influxpush/parser.go

```diff
@@ -0,0 +1,188 @@
+	precision := qp.Get("precision")
+	if precision == "" {
+		precision = "ns"
+	}
+
+	if !models.ValidPrecision(precision) {
+		return nil, 0, fmt.Errorf("precision supplied is not valid: %s", precision)
+	}
+
+	encoding := r.Header.Get("Content-Encoding")
+	reader, err := batchReadCloser(r.Body, encoding, int64(maxSize))
+	if err != nil {
+		return nil, 0, fmt.Errorf("gzip compression error: %w", err)
+	}
+	data, err := io.ReadAll(reader)
+	dataLen := len(data) // In case if something is read despite an error.
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("can't read body: %s", err)
+	}
+
+	err = reader.Close()
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("problem reading body: %s", err)
+	}
+
+	points, err := models.ParsePointsWithPrecision(data, time.Now().UTC(), precision)
+	if err != nil {
+		return nil, dataLen, fmt.Errorf("can't parse points: %s", err)
+	}
+	ts, err := writeRequestFromInfluxPoints(points)
+	return ts, dataLen, err
+}
+
+func writeRequestFromInfluxPoints(points []models.Point) ([]mimirpb.PreallocTimeseries, error) {
+	// Technically the same series should not be repeated. We should put all the samples for
+	// a series in single client.Timeseries. Having said that doing it is not very optimal and the
+	// occurrence of multiple timestamps for the same series is rare. Only reason I see it happening is
+	// for backfilling and this is not the API for that. Keeping that in mind, we are going to create a new
+	// client.Timeseries for each sample.
+
+	returnTs := mimirpb.PreallocTimeseriesSliceFromPool()[:0]
+	if cap(returnTs) < len(points) {
+		returnTs = make([]mimirpb.PreallocTimeseries, 0, len(points))
+	}
+	for _, pt := range points {
+		var err error
+		returnTs, err = influxPointToTimeseries(pt, returnTs)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return returnTs, nil
+}
+
+// Points to Prometheus is heavily inspired from https://github.com/prometheus/influxdb_exporter/blob/a1dc16ad596a990d8854545ea39a57a99a3c7c43/main.go#L148-L211
+func influxPointToTimeseries(pt models.Point, returnTs []mimirpb.PreallocTimeseries) ([]mimirpb.PreallocTimeseries, error) {
+
+	fields, err := pt.Fields()
+	if err != nil {
+		return returnTs, fmt.Errorf("can't get fields from point: %w", err)
+	}
+	for field, v := range fields {
+		var value float64
+		switch v := v.(type) {
+		case float64:
+			value = v
+		case int64:
+			value = float64(v)
+		case bool:
+			if v {
+				value = 1
+			} else {
+				value = 0
+			}
+		default:
+			continue
+		}
+
+		name := string(replaceInvalidChars(pt.Name()))
+		if field != "value" {
+			// If the field name is not "value" then we append it to the name, fixing chars as we go
+			name += "_" + string(replaceInvalidChars([]byte(field)))
+		}
+
+		tags := pt.Tags()
+		lbls := make([]mimirpb.LabelAdapter, 0, len(tags)+2) // An additional one for __name__, and one for internal label
+		lbls = append(lbls, mimirpb.LabelAdapter{
+			Name:  labels.MetricName,
+			Value: name,
+		})
```

Comment:
```suggestion
		name := replaceInvalidChars(pt.Name())
		if field != "value" {
			// If the field name is not "value" then we append it to the name, fixing chars as we go
			name = append([]byte{'_'}, name...)
		}

		tags := pt.Tags()
		lbls := make([]mimirpb.LabelAdapter, 0, len(tags)+2) // An additional one for __name__, and one for internal label
		lbls = append(lbls, mimirpb.LabelAdapter{
			Name:  labels.MetricName,
			Value: yoloString(name),
		})
```

Reply:
This didn't quite work. If the initial results of `pt.Name()` was modified (e.g. with a `_` byte prepended) then subsequent calls to `pt.Name()` could return incorrect results.

I've rejigged it to make a copy of the slice returned by `pt.Name()` and then append as necessary to that, keeping it as a `[]byte` rather than converting to a string.